bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+                to,
+                msg: Request {
+                    id: request_id,
+                    type_: RequestType::GetShare { rack_uuid },
+                }
+                .into(),
+            })
+            .collect();
+
+        (request_id, Output { persist: false, envelopes, api_output: None })
+    }
+
+    /// We received a share
+    ///
+    /// Verify that the share is valid and recompute the secret if there are
+    /// enough shares.
+    pub fn on_share(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        share: Vec<u8>,
+        rack_secret_expiry: Ticks,
+    ) -> Output {
+        match self {
+            RackSecretState::Empty => ApiError::UnexpectedResponse {
+                from,
+                state: "-",
+                request_id,
+                msg: "share",
+            }
+            .into(),
+            RackSecretState::Retrieving {
+                from: from_all,
+                shares,
+                threshold,
+                share_digests,
+                ..
+            } => {
+                // Ignore the share if we already have one from this peer
+                if from_all.contains(&from) {
+                    return Output::none();
+                }
+
+                // Compute and validate hash of the received key share
+                if let Err(api_error) =
+                    Self::validate_share(&from, &share, &share_digests)
+                {
+                    return api_error.into();
+                }
+
+                // Keep track of our new valid share
+                shares.insert(share);
+                from_all.insert(from);
+
+                // If we have enough shares, try to recompute the rack secret
+                // If we fail here, this is a major problem so we report it and
+                // clear our rack secret state, as we won't be able to succeed.
+                // This error should be effectively impossible to ever hit.
+                if shares.0.len() == *threshold {
+                    match RackSecret::combine_shares(&shares.0) {
+                        Ok(secret) => {
+                            *self = RackSecretState::Computed {
+                                secret: secret.clone(),
+                                expiry: rack_secret_expiry,
+                            };
+                            ApiOutput::RackSecret(secret).into()
+                        }
+                        Err(_) => {
+                            *self = RackSecretState::Empty;
+                            ApiError::FailedToReconstructRackSecret.into()
+                        }
+                    }
+                } else {
+                    // We don't have enough shares
+                    Output::none()
+                }
+            }
+            RackSecretState::Computed { .. } => {
+                // This is just a late share response. We've already informed
```

Comment:
Is there a reason to store this data, just in case?

Reply:
No longer relevant, but no. We only need a threshold to recompute the secret. Storing extra shares is unnecessary and leaks information.